### ISSUE TITLE
Python lsp: Pyright doesn't good

### ISSUE THREAD

**Issue by lctzz540** [zed-industries/zed#7296]
### Check for existing issues

- [X] Completed

### Describe the feature

it should be pylsp. The pyright has issues about recognize libraries installed

**Comment by JosephTLyons** [comment]
Yeah, Pyright has been a big source of pain for us in the Zed-Python intersection. I think we should def swap out to something better. I've not heard of `pylsp`, but we should def add it to the list of ones to check out:

- [`pylsp`](https://github.com/python-lsp/python-lsp-server)
- [`pylyzer`](https://github.com/mtshiba/pylyzer)
- [`astral.sh`](https://astral.sh) will hopefully build a language server in Rust

We are working on language plugins at the moment, so, in the not-so-distant future, someone can just add these servers as plugins, but I do think we should swap out our provided one for a bettter one.

**Comment by plattenschieber** [comment]
> Yeah, `pyright` has been a big source of pain for us in the Zed-Python intersection. I think we should def swap out to something better. I've not heard of `pylsp`, but we should def add it to the list of ones to check out:
> 
> * [`pylsp`](https://github.com/python-lsp/python-lsp-server)
> * [`pylyzer`](https://github.com/mtshiba/pylyzer)
> * [`astral.sh`](https://astral.sh) will hopefully build a language server in Rust
> 
> We are working on language plugins at the moment, so, in the not-so-distant future, someone can just add these servers as plugins, but I do think we should swap out our provided one for a bettter one.

Indeed they did build a language server protocol for Ruff already [here](https://github.com/astral-sh/ruff-lsp) @JosephTLyons 🔥

**Comment by JosephTLyons** [comment]
> > Yeah, `pyright` has been a big source of pain for us in the Zed-Python intersection. I think we should def swap out to something better. I've not heard of `pylsp`, but we should def add it to the list of ones to check out:
> > 
> > * [`pylsp`](https://github.com/python-lsp/python-lsp-server)
> > * [`pylyzer`](https://github.com/mtshiba/pylyzer)
> > * [`astral.sh`](https://astral.sh) will hopefully build a language server in Rust
> > 
> > We are working on language plugins at the moment, so, in the not-so-distant future, someone can just add these servers as plugins, but I do think we should swap out our provided one for a bettter one.
> 
> Indeed they did build a language server protocol for Ruff already [`ruff-lsp`](https://github.com/astral-sh/ruff-lsp) @JosephTLyons 🔥

AFAICT, `ruff-lsp` only support formatting and linting as of right now, as they haven't gotten to building things out like `go to def` and auto-completion, but I read they will be doing that at some point.

**Comment by liebkne** [comment]
https://github.com/pappasam/jedi-language-server

**Comment by JacobGoldenArt** [comment]
Yeah. I love Zed but when I open a Python project it's just strange. I will activate my .venv in my project directory and the script will run but when you look at the python file the imports are all red and can't be resolved. 
<img width="901" alt="Screenshot 2024-03-02 at 7 58 28 PM" src="https://github.com/zed-industries/zed/assets/126922704/3b8213f1-b52a-470c-af1e-da7c73c2772f">

**Comment by JacobGoldenArt** [comment]
See that python-dotenv is installed, along with pydantic in my env - (pip list) but I guess pyright is not checking for packages in the project-directory/.venv ???

**Comment by failable** [comment]
You need this file.

```
$ cat pyrightconfig.json
{
  "venvPath": ".",
  "venv": ".venv"
}
```

**Comment by Moshyfawn** [comment]
There's an issue #8541 about resolving modules with Pyright.

**Comment by newtome8888** [comment]
I see this problem last for months, but there's still no really solution?

**Comment by JosephTLyons** [comment]
> You need this file.
> 
> ```
> $ cat pyrightconfig.json
> {
>   "venvPath": ".",
>   "venv": ".venv"
> }
> ```

This should actually solve the problem - for some reason, on our documentation migration, we lost our python documentionat that mentions this config file. I've quickly dumped the old python docs contents into the `configuration` section of our website, [here](https://zed.dev/docs/python). I need to bring back more documentation, so I will organize this better later on.

As a side note, I want to say that we are going to doing more for Python here shortly - we recognize just how important this language is, being one of the top most-used languages, and we want to improve things here.

**Comment by collinsinclair** [comment]
To add to the corpus of Pyright issues, in `is_24hour = models.BooleanField(default=False)`, `False` is red-underlined for
> Argument of type "Literal[False]" cannot be assigned to parameter "default" of type "type[NOT_PROVIDED]" in function "\__init__" Type "Literal[False]" cannot be assigned to type "type[NOT_PROVIDED]"

Additionally, in `def __str__(self):`, `__str__` is red-underlined for

> Method "\__str__" overrides class "Model" in an incompatible manner Return type mismatch: base method returns type "str", override returns type "CharField" "CharField" is incompatible with "str"

I have done no custom configuration. I believe these shouldn't be showing issues as they work as expected and are standard practices.

**Comment by r-cha** [comment]
> I tried it, but no luck. I use poetry in my python project, does it matter?

@newtome8888 run `poetry env info` to find the path to your virtualenv, then add that to your pyrightconfig. Worth noting, this path will be different per machine, so don't add `pyrightconfig.json` to any version control. Mine looks something like this:

```json
{
  "venvPath": "/Users/username/Library/Caches/pypoetry/virtualenvs",
  "venv": "projectname-p0LYZhue-py3.11"
}
```

**Comment by JosephTLyons** [comment]
> > I tried it, but no luck. I use poetry in my python project, does it matter?
> 
> @newtome8888 run `poetry env info` to find the path to your virtualenv, then add that to your pyrightconfig. Worth noting, this path will be different per machine, so don't add `pyrightconfig.json` to any version control. Mine looks something like this:
> 
> ```json
> {
>   "venvPath": "/Users/username/Library/Caches/pypoetry/virtualenvs",
>   "venv": "projectname-p0LYZhue-py3.11"
> }
> ```

Just to mention, there is a setting in poetry tells poetry to install the venv within the project:

- https://python-poetry.org/docs/configuration#virtualenvsin-project

Which makes things a bit more portable, if you are running the project on many different machines...

**Comment by failable** [comment]
Using pyright is literally painful.

**Comment by myusuf3** [comment]
> To add to the corpus of Pyright issues, in `is_24hour = models.BooleanField(default=False)`, `False` is red-underlined for
> 
> > Argument of type "Literal[False]" cannot be assigned to parameter "default" of type "type[NOT_PROVIDED]" in function "__init__" Type "Literal[False]" cannot be assigned to type "type[NOT_PROVIDED]"
> 
> Additionally, in `def __str__(self):`, `__str__` is red-underlined for
> 
> > Method "__str__" overrides class "Model" in an incompatible manner Return type mismatch: base method returns type "str", override returns type "CharField" "CharField" is incompatible with "str"
> 
> I have done no custom configuration. I believe these shouldn't be showing issues as they work as expected and are standard practices.

I am running into the same issue. it's hurting more than its helping.

**Comment by strangemonad** [comment]
@JosephTLyons I'd like to offer a counter proposal - pyright is actually really good... assuming it gets pointed to the right thing. Expanding on https://github.com/zed-industries/zed/issues/7808#issuecomment-2012710492

1) Putting aside the config / invocation of pyright for a moment, I believe the approach to how the type system was implemented in the pyright type-checker leads to the best editor experience. There's a few subtle choices e.g. how bi-directly type inference within a method with partial type annotations are handled compared to mypy that make a much better dev experience while editing python code. It's also faster than mypy and more complete than any other option currently available in my experience.

2) robustly invoking pyright is the key. I think the docs are clunky but the most robust approach effectively boils down to the following

```
pyright --pythonpath $ABS_PATH_TO_PYTHON --project $ABS_PATH_TO_PROJECT_ROOT
```

For example, if I have a workspace pointed at the following folders
```
api
    - pyproject.toml # poetry project
    - src/
    - tests/
backend
    - pyproject.toml # poetry project
    - src/
    - tests/
frontend
    - yarn / node...
```

To get pyright working for `backend` I would do `cd backend && pyright --pythonpath $(poetry env info --executable) --project $PWD`

Why not use the `venv` `venvPath` options? If you scrutinize the pyright docs, they mention that this isn't really scalable because it's hard to enforce that every dev creates a venv in the same location by default. The `pythonpath` option is how vscode invokes the pyright lsp.

Why not use the execution environments option?  In practice, while you can have a pyright config file in the project root, you might want to open multiple different folders in a single editor workspace.

**Comment by Yevgnen** [comment]
@strangemonad Can you share the "actually really good" parts? I don't really get it...

**Comment by strangemonad** [comment]
@Yevgnen sure, I can give it a shot. I think it's helpful to break this up into a few categories:

*everything below being said I do think the ease of use of pyright could be improved. I'd also like to remind folks that even though pyright comes out of Microsoft, for reasons I don't understand it's primarily the work of only a single person - including documentation and triaging of bugs*

1. choice of LSP
2. LSP vs type checker vs linter
3. Speed
4. PyRight's type system

**1**
I'm still very much in favor of user choice of LSP. Folks should be able to pick what's currently best at their current project's needs. I primarily do python work in a single editor workspace that contains many python projects in various workspace directors (e.g. VSCode's add folder to workspace). Most of these are CLIs, services, or ML model pipelines. This is probably pretty different than the needs of someone who works on, say, a single python open source library at a time.

**2**
There might be some LSPs that work separately from a downstream type checker. Personally, I like a couple invariants
    a) I want anything that would fail a type checking command or lint command to show up inline (red / yellow squiggles) as I'm writing the code.
    b) I don't find it personally acceptable if symbol completion suggests something that's not a valid symbol and I end up getting method not found or what not when I run the code
I practice, that pretty much means that whatever the LSP is calling out to needs to be equivalent to the same type checker implementation. E.g. if someone were able to make a better python LSP front end that called out to pyright + ruff and combined results, that would be great (assuming they could get it working reliably which would be no small feat given the current state of the world).

**Speed**
Once PyRight is up and running, responses and updates feel faster to me than other approaches. I don't have empirical evidence for this, just what it feels like to me.

**Type system**
This is probably the most nuanced but I think it makes a big difference, especially around producing false-negatives and suggesting type hint corrections while typing. Some back context: As type hints have been gradually added to Python it's been done differently than other languages that have added progressive optional typing (e.g. typescript). Effectively, when taken together, the various python type hint PEPs don't really define a type system, but rather provide the guidelines and specifications for a possible type system. Rather than talk about the "python type system" it would probably be more correct to say that MyPy, Pyright, pytypre, pyre have each implemented "a python type system" with different design constraints to suit specific needs. I'd consider MyPy and PyRight to be the 2 most general implementations. Pytype has made choices that make it more applicable to Google's large (and older) python code base. Pyre has learned a lot from flowtype but is aimed a bit more at finding security vulnerabilities.

Putting aside the speed difference between pyright and mypy, there's a bunch of subtle design differences that might seem insignificant but I think end up mattering a lot when editing code. To boil it down to just 2:

1) How missing type hints are handled
2) Limited-scope type inference.

MyPy interprets the PEP quite literally. e.g. if a function has no return type hint, pyright treats it as if it had the `Any` return type. Pyright, however treat's it as "untyped" and then treats this "Untyped" fictitious type as being equivalent to `Any`.

Within bounded scopes (e.g. the scope of a method) pyright implements bi-direction type checking (in the style of Pierce) e.g. even though `def foo(a, b):` has no type hints, if the first line is `a + b` we know we at the very least expect `a` to implement `__add__` all these bits of information within a method to either catch errors that can't be caught if you don't do type inference or suggest type hints for parameters or the return type.

**Comment by geoqiao** [comment]
Is it possible to add a global setting in Zed like pylance in vscode? 
It's really a pain to have to manually add` pyrightconfig.json` every time  creating a Python project.https://github.com/zed-industries/zed/issues/10587

**Comment by didibear** [comment]
@geoqiao pyright can be configured in the `pyproject.toml` [as documented here](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#pyright-configuration), for example like this:

```toml
[tool.pyright]
venv = ".venv"
venvPath = "."
```

**Comment by geoqiao** [comment]
> @geoqiao pyright can be configured in the `pyproject.toml` directly [as documented here](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#pyright-configuration), for example like this:
> 
> 
> 
> ```toml
> 
> [tool.pyright]
> 
> venv = ".venv"
> 
> venvPath = "."
> 
> ```

It still needs to be manually configured.

**Comment by vladimir-sama** [comment]
```
item_icon : Optional[FileStorage] = request.files.get('item_icon')
if item_icon:
  file_name : str = item_id + '.' + item_icon.filename.rsplit('.', 1)[1].lower()
```
`Pyright: "rsplit" is not a known attribute of "None"`
Good typing system? It doesn't even do TypedDict correctly

So, will there be an option system or extensions soon? I'd like Pylance but I'm not sure if it can be an extension. I've tried MyPy, but cannot remember if it was good

**Comment by injust** [comment]
> ```
> item_icon : Optional[FileStorage] = request.files.get('item_icon')
> if item_icon:
>   file_name : str = item_id + '.' + item_icon.filename.rsplit('.', 1)[1].lower()
> ```
> 
> `Pyright: "rsplit" is not a known attribute of "None"` Good typing system? It doesn't even do TypedDict correctly

Off-topic, but @vladimir-sama your example is simply user error: `FileStorage.filename` can be `None`. See https://github.com/pallets/werkzeug/blob/3.0.3/src/werkzeug/datastructures/file_storage.pyi#L12.

**Comment by failable** [comment]
Does Zed support to config pyright to make it work with workspaces?

**Comment by jaanli** [comment]
yes, i am also having this issue :(

**Comment by yash-desai-naik** [comment]
Just simply open the project form the terminal with env activated!!!

I faced the same issue when opening project directory directly in zed, pyright started to yell "missing libraries".

# Solution:

1. Open terminal
2. `cd YOUR_PROJECT_DIRECTORY`
3. `source .venv/bin/activate`
4. `zed .`

now it should work smoothly

**Comment by andre0xFF** [comment]
> Just simply open the project form the terminal with env activated!!!
> 
> I faced the same issue when opening project directory directly in zed, pyright started to yell "missing libraries".
> 
> # Solution:
> 1. Open terminal
> 2. `cd YOUR_PROJECT_DIRECTORY`
> 3. `source .venv/bin/activate`
> 4. `zed .`
> 
> now it should work smoothly

doesn't work. activated the environment with poetry and still get unresolved imports.

**Comment by geoqiao** [comment]
> > Just simply open the project form the terminal with env activated!!!
> 
> > 
> 
> > I faced the same issue when opening project directory directly in zed, pyright started to yell "missing libraries".
> 
> > 
> 
> > # Solution:
> 
> > 1. Open terminal
> 
> > 2. `cd YOUR_PROJECT_DIRECTORY`
> 
> > 3. `source .venv/bin/activate`
> 
> > 4. `zed .`
> 
> > 
> 
> > now it should work smoothly
> 
> 
> 
> doesn't work. activated the environment with poetry and still get unresolved imports.

This is a problem with pyright config, which has nothing to do with whether to activate the python virtual environment.

We need a better way to make pyright work without generating pyryghtconfig.json every time opening a new project.

**Comment by Act0r1** [comment]
> You need this file.
> 
> ```
> $ cat pyrightconfig.json
> {
>   "venvPath": ".",
>   "venv": ".venv"
> }
> ```

it is not solve problem tbh

**Comment by KraXen72** [comment]
i'm using https://github.com/DetachHead/basedpyright in vscode and it's been great so far at replacing normal pyright or pylance. could be worth a shot.

**Comment by bluekeybo** [comment]
@JosephTLyons when will this be resolved? The python devs are suffering, in fact Zed is unusable until this is resolved. Seeing unresolved imports is a pain. Thanks!

**Comment by FilipeBisinella** [comment]
I made a PR to allow for pyright settings, which could help with this.

https://github.com/zed-industries/zed/pull/14265

**Comment by jdoss** [comment]
Is switched to https://github.com/DetachHead/basedpyright a bit a go in VSCode and it's been great. It should be a drop in replacement for pyright. ™️ 

Maybe make it so people can can choose between the two? One thing that is great about basedpyright is it is published as a pypi package so you can set the version you want inside of your python project.

https://docs.basedpyright.com/#/?id=published-as-a-pypi-package-no-nodejs-required

**Comment by m1guer** [comment]
> placement

Hi, i have done one, but i cant test on linux(maybe a bug or skill issue), if you wanna test:
https://github.com/m1guer/basedpyright-zed

**Comment by failable** [comment]
@m1guer How to use? I select the directory on "Install Dev Extension" and nothing happens.

**Comment by m1guer** [comment]
i fixed some bugs, maybe works now.
try this:
```
{
  "languages": {
    "Python": {
     "language_servers":["!pyright","basedpyright"]
    }
  }
}
```

**Comment by failable** [comment]
@m1guer Just tried and does not work even with the setting. The extension does not appear in the Extension list.

**Comment by m1guer** [comment]
can open your logs?

**Comment by failable** [comment]
Here you go

```
a9c2705b-6230-4d52-93ce-787456177e81
2024-07-17T23:29:18.09519Z [INFO] compiling Rust extension /Users/user/Downloads/basedpyright-zed
2024-07-17T23:29:18.210935Z [INFO] compiling rust extension /Users/user/Downloads/basedpyright-zed
2024-07-17T23:29:18.389182Z [INFO] checking out python parser
2024-07-17T23:29:18.399757Z [ERROR] failed to compile grammar 'python'

Caused by:
    grammar directory '/Users/user/Downloads/basedpyright-zed/grammars/python' already exists, but is not a git clone of 'https://github.com/tree-sitter/tree-sitter-python'
```

**Comment by m1guer** [comment]
try delete the grammar folder in the extension dir and install with pip install badedpyright

**Comment by failable** [comment]
@m1guer Works like a charm.

**Comment by m1guer** [comment]
:)
has something to do.
like do all the schemas for the language.

**Comment by Greg-Hamel** [comment]
@m1guer Thanks for bringing `basedpyright` to Zed!

For those curious, the extension is under review at https://github.com/zed-industries/extensions/pull/1080

**Comment by code-yeongyu** [comment]
Since `pyright` and `basedpyright` don't support formatting, I agree with implementing pylsp as default. Here's why [`pylsp`](https://github.com/python-lsp/python-lsp-server) could be a great default:

1. Extensibility:
   - [Supports various plugins](https://github.com/python-lsp/python-lsp-server?tab=readme-ov-file#3rd-party-plugins)
     - Could potentially serve as an adapter for Python environment setup in Zed Editor
2. Comprehensive tooling via plugins:
   - [`mypy`](https://github.com/python/mypy) plugin for type checking
   - [`ruff`](https://github.com/astral-sh/ruff) plugin for formatting and linting
   - Built-in autocompletion powered by [`rope`](https://github.com/python-rope/rope)

I was thinking that pylsp with mypy and ruff plugins might work well as a default preset. This would facilitate setting up a professional Python development environment.

**This preset could allow users to access autocompletion, linting, type checking, and formatting simply by choosing to use pylsp.**

Other LSP implementations (pyright, basedpyright, ruff) could be offered as optional extensions for flexibility, giving users more choices if they have specific preferences.

I'd appreciate any thoughts on this approach.

**Comment by code-yeongyu** [comment]
> > I agree with implementing pylsp as default.
> > ...
> > I'd appreciate any thoughts on this approach.
> 
> Give it a shot as an extension - [zed-industries/extensions@`main`/AUTHORING_EXTENSIONS.md](https://github.com/zed-industries/extensions/blob/main/AUTHORING_EXTENSIONS.md?rgh-link-date=2024-07-23T15%3A01%3A46Z). I've been tempted to try it once I finish some other work. Getting an extension out there will help people figure out quickly what's best and how we could incorporate it more deeply.

Thank you for mentioning that. As a passionate Python developer and a Rust fanboy, this seems like a great first step to contribute to Zed.

**Comment by rgbkrk** [comment]
Sweet, this is probably a good starting point: https://github.com/zed-industries/zed/pull/14198/files

H/T @osiewicz

**Comment by Greg-Hamel** [comment]
@code-yeongyu Please link your PR to this issue, when you get to it, so we can give it a shot. I'd be very happy to test it for my own use case.

I agree with the opinion that an all-in-one lsp + formatter would be well received.

I currently run `black` on the side on save at the moment.

**Comment by m1guer** [comment]
> @m1guer Thanks for bringing `basedpyright` to Zed!
> 
> For those curious, the extension is under review at [zed-industries/extensions#1080](https://github.com/zed-industries/extensions/pull/1080)

your welcome, im just waiting for the review approvement. i fixed all the bugs related to the schema.

**Comment by hirako2000** [comment]
neither `pyright.json` and `myproject.toml` with the pyright setting worked for me.
In case someone faces the same fate this fixed it:

```bash
mkdir .zed # in the root folder of the project
touch .zed/settings.json
```

in that `settings.json`, having installed ruff as linter.
```json
{
  "tools": {
    "ruff": {
      "command": "ruff",
      "args": ["--format", "json", "--stdin-filename", "$FILENAME", "-"],
      "rootPatterns": ["pyproject.toml", "setup.cfg", ".git"],
      "filePatterns": ["*.py"],
      "runOnSave": true,
      "errorPatterns": [
        {
          "regexp": "^(.*?):(\\d+):(\\d+): (.*)$",
          "file": 1,
          "line": 2,
          "column": 3,
          "message": 4
        }
      ]
    }
  }
}

if using another linter, adjust accordingly.
```

**Comment by chrisemke** [comment]
Is there any chance of support for the mypy daemon?
I currently use ruff and mypy together in vscodium, it would be nice to have the mypy information inside the editor as well using dmypy

https://mypy.readthedocs.io/en/stable/mypy_daemon.html

**Comment by v3ss0n** [comment]
Any updates here? current pyright in Zed is down right disapponting . 
Need BasedPyRIght + Ruff + MyPY.

**Comment by m1guer** [comment]
ruff and basedpyright are on extensions.

**Comment by rgbkrk** [comment]
I don't want people to be burdened by too much choice but I put together an extension using `python-lsp-server` you can try by installing as a dev extension: https://github.com/rgbkrk/python-lsp-zed-extension

Thus far I'm enjoying it and it feels like one I should submit to the extensions repo. `python-lsp-server` has mypy support but I've yet to see it working in practice. Happy to hack with folks on it! Find me in #repl on the zed channel.

**Comment by bersace** [comment]
Wow, pyright eats 3.5G of RAM after a few days. Never had this with jedi.

**Comment by bersace** [comment]
> Thus far I'm enjoying it and it feels like one I should submit to the extensions repo. `python-lsp-server` has mypy support but I've yet to see it working in practice. Happy to hack with folks on it!

How to disable pyright and move to pylsp ?

**Comment by a-kachurin** [comment]
> You can _now_ search for "Python LSP" in extensions to enable it.
> 
> ![image](https://private-user-images.githubusercontent.com/836375/360977815-f2c5a4fd-e3c8-40a3-a2fa-923fb6d7ba86.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ1MDA5OTEsIm5iZiI6MTcyNDUwMDY5MSwicGF0aCI6Ii84MzYzNzUvMzYwOTc3ODE1LWYyYzVhNGZkLWUzYzgtNDBhMy1hMmZhLTkyM2ZiNmQ3YmE4Ni5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyNFQxMTU4MTFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYWI2ZjVjYjdkMTBlYzgxYjkxNTk2YjMyNDE2MGIyMzI1OWRmZjU5MzQ3ZDA1MTRkZmUyNzZiM2UzMTc5YWRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Dsbjr4IOVdWOYrQ--OMnJT9KJ3b_mjZvFFyC5PV3IjM)
> 
> Note that you will need to [install the python lsp server directly first](https://github.com/python-lsp/python-lsp-server).
> 
> Read more in the [`python-lsp-zed-extension` README](https://github.com/rgbkrk/python-lsp-zed-extension/blob/main/README.md) and file issues there. I'm still figuring out what else I can enable. I would ❤️ if someone gets mypy configuration working with it.

Is the "Go to Definition" feature implemented in your plugin? I would like to try Zed on Win11. Could you please provide more detailed instructions on how to set up the plugin?

**Comment by failable** [comment]
Has anyone successfully started pylsp and been able to use the code completion feature? Even though I installed `python-lsp-server`, I still can't use any of its functions.

**Comment by vrslev** [comment]
@failable You should also have `pylsp` installed—then it works

**Comment by failable** [comment]
@vrslev 

 > Even though I installed python-lsp-server

**Comment by vrslev** [comment]
Do you have it in PATH?

**Comment by fabioz** [comment]
Question (I can open a different issue if needed, but I thought asking here is Ok, if not, please let me know).

I have my own language server that I do for Python (https://marketplace.visualstudio.com/items?itemName=fabioz.vscode-pydev), how can I stop using the builtin one to use my own? 

-- i.e.: if I register another Python extension (as in https://github.com/rgbkrk/python-lsp-zed-extension/), will it stop using the builtin one, which is giving me pyright errors?

**Comment by Herraj** [comment]
Looks like `pylsp` was added as of today: https://github.com/zed-industries/zed/pull/20358. Also mentioned in the release notes!

**Comment by JaagupAverin** [comment]
> Since `pyright` and `basedpyright` don't support formatting, I agree with implementing pylsp as default
> ...
> I'd appreciate any thoughts on this approach.

My only gripe with pylsp is that, as of today, basedpyright is a more fleshed out type checker with more features such as inlay hints, and a very active developer who's pushing patches on a weekly basis, whereas pylsp seems to be a bit stale and features such as inlay hints are [still missing](https://github.com/python-lsp/python-lsp-server/issues/397).

Furthermore from my perspective having the formatter (Ruff) in parallel to the type checker is a more flexible experience than having the formatter hierarchically under the the type checker (as a plugin). But I am not well acquainted with the technicalities of this topic so perhaps there is a performance/scaling consideration that I'm not aware of.

**Comment by ryukinix** [comment]
There is any simple way to replace pyright with mypy? Pyright being the default is kinda a deal breaker for me.

**Comment by osiewicz** [comment]
You can use mypy with pylsp that's the other language server provided out of the box.

**Comment by smparekh** [comment]
for macOS using pylsp with ruff, here is my setup:
```
...
  "languages": {
    "Python": {
      "language_servers": ["pylsp", "!pyright", "..."],
      "preferred_line_length": 120
    }
  },
  "lsp": {
    "pylsp": {
      "binary": {
        "path": "/Users/<USERNAME>/.local/bin/pylsp"
      },
      "settings": {
        "plugins": {
          "ruff": {
            "enabled": true,
            "formatEnabled": true,
            "lineLength": 120,
            "targetVersion": "py311"
          }
        }
      }
    },
...
```

Install python-lsp-ruff:
```
pipx install python-lsp-ruff --include-deps
```

**Comment by g0di** [comment]
> Furthermore from my perspective having the formatter (Ruff) in parallel to the LSP is a more flexible experience than having the formatter hierarchically under the the LSP (as a plugin). But I am not well acquainted with the technicalities of this topic so perhaps there is a performance/scaling consideration that I'm not aware of.

I can rely on that. I've give a try to `python-lsp-ruff` plugin to use only `pylsp` as my one and only language server in Zed for python and it appeared that formatting (and code fixes) were lot more slower (more than a second) compared to using directly the `ruff` lsp (instantaneous) next to `pylsp` or `pyright` for handling the rest. I don't know if I'm the only one noticing this.

By the way, I actually have to use `pyright`, `pylsp` and `ruff` servers next to each others in my Zed config because using only `pylsp` and `ruff` does not provide any autocompletion anymore... Not sure why. So `pyright` gives me autocompletion (and more), `pylsp` gives me `mypy` type checking (through plugin) and `ruff` gives me linting and formatting.

Not sure if this is the best config but it works as intended. Just my 2 cents experience.
